central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+				Protocol:           val.Protocol,
+				ProcessIndicatorId: indicatorID,
+				Process:            processInfo,
+				Closed:             val.CloseTimestamp != nil,
+				CloseTimestamp:     val.CloseTimestamp,
+			}
+
+			plopObjects[i] = newPLOP
+		}
+	}
+
+	// Now save actual PLOP objects
+	err = ds.storage.UpsertMany(ctx, plopObjects)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ds *datastoreImpl) GetProcessListeningOnPort(
+	ctx context.Context,
+	deployment string,
+) (
+	processesListeningOnPorts []*storage.ProcessListeningOnPort, err error,
+) {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
+		return nil, err
+	} else if !ok {
+		return nil, sac.ErrResourceAccessDenied
+	}
+
+	processesListeningOnPorts, err = ds.storage.GetProcessListeningOnPort(ctx, deployment)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return processesListeningOnPorts, nil
+}
+
+func (ds *datastoreImpl) WalkAll(ctx context.Context, fn WalkFn) error {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	return ds.storage.Walk(ctx, fn)
+}
+
+func (ds *datastoreImpl) RemovePLOP(ctx context.Context, ids []string) error {
+	if ok, err := plopSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	return ds.removePLOP(ctx, ids)
+}
+
+func (ds *datastoreImpl) removePLOP(ctx context.Context, ids []string) error {
+	log.Infof("Deleting PLOP records")
+
+	if len(ids) == 0 {
+		return nil
+	}
+	if err := ds.storage.DeleteMany(ctx, ids); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// fetchExistingPLOPs: Query already existing PLOP objects belonging to the
+// specified process indicators.
+//
+// XXX: This function queries all PLOP, no matter if they are matching port +
+// protocol we've got or not. This means potentially dangerous corner cases
+// when one process listenst to a huge number of ports. To address it we could
+// introduce filtering by port and protocol to the query, and even without
+// extra indices PostgreSQL will be able to do it relatively efficiently using
+// bitmap scan.
+func (ds *datastoreImpl) fetchExistingPLOPs(
+	ctx context.Context,
+	indicatorIds []string,
+	portProcesses ...*storage.ProcessListeningOnPortFromSensor,
+) (map[string]*storage.ProcessListeningOnPortStorage, error) {
+
+	var existingPLOPMap = map[string]*storage.ProcessListeningOnPortStorage{}
+
+	if len(indicatorIds) == 0 {
+		return nil, nil
+	}
+
+	// If no corresponding processes found, we can't verify if the PLOP
+	// object is opening/closing an existing one. Collect existingPLOPMap
+	// only if there are some matching indicators.
+	existingPLOP, err := ds.storage.GetByQuery(ctx, search.NewQueryBuilder().
+		AddStrings(search.ProcessID, indicatorIds...).ProtoQuery())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, val := range existingPLOP {
+		key := fmt.Sprintf("%d %d %s",
```

Comment:
same comment as earlier on the spaces.  Additionally we should probably create a helper function to create the key.

Reply:
See above response to similar comment.